server/src/test/java/org/elasticsearch/action/support/SubscribableListenerTests.java

```diff
@@ -604,4 +606,87 @@ public void testAndThenAcceptFailure() {
     }
+
+    public void testJavaDocExample() {
+        safeAwait(SubscribableListener.<Boolean>newForked(l -> exampleAsyncMethod(randomIdentifier(), List.of(), l)));
+    }
+
+    private void exampleAsyncMethod(String request, List<Long> items, ActionListener<Boolean> finalListener) {
+        SubscribableListener
+
+            // Start the chain and run the first step by creating a SubscribableListener using newForked():
+            .<String>newForked(l -> firstAsyncStep(request, l))
+
+            // Run a second step when the first step completes using andThen(); if the first step fails then the exception falls through to
+            // the end without executing the intervening steps.
+            .<Integer>andThen((l, firstStepResult) -> secondAsyncStep(request, firstStepResult, l))
+
+            // Run another step when the second step completes with another andThen() call; as above this only runs if the first two steps
+            // succeed.
+            .<Boolean>andThen((l, secondStepResult) -> {
+                if (condition) {
+                    // Steps are exception-safe: an exception thrown here will be passed to the listener rather than escaping to the
+                    // caller.
+                    throw new IOException("failure");
+                }
+
+                // Steps can fan out to multiple subsidiary async actions using utilities like RefCountingListener.
+                final var result = new AtomicBoolean();
+                try (var listeners = new RefCountingListener(l.map(v -> result.get()))) {
+                    for (final var item : items) {
+                        thirdAsyncStep(secondStepResult, item, listeners.acquire());
+                    }
+                }
+            })
+
+            // Synchronous (non-forking) steps which do not return a result can be expressed using andThenAccept() with a consumer:
+            .andThenAccept(thirdStepResult -> {
+                if (condition) {
+                    // andThenAccept() is also exception-safe
+                    throw new ElasticsearchException("some other problem");
+                }
+                consumeThirdStepResult(thirdStepResult);
+            })
+
+            // Synchronous (non-forking) steps which do return a result can be expressed using andThenApply() with a function:
+            .andThenApply(voidFromStep4 -> {
+                if (condition) {
+                    // andThenApply() is also exception-safe
+                    throw new IllegalArgumentException("failure");
+                }
+                return computeFifthStepResult();
+            })
+
+            // To complete the chain, add the outer listener which will be completed with the result of the previous step if all steps were
+            // successful, or the exception if any step failed.
+            .addListener(finalListener);
+    }
+
+    private static final boolean condition = false;
```

Comment:
since we're testing and you've written the error handling already, should we also test the throwing scenario?

Reply:
Ah this is all just here to make sure the example in the docs compiles. The behavioural tests are elsewhere in this suite. See 8f70831a5ad.